node/service/src/lib.rs

```diff
@@ -428,28 +513,78 @@ pub fn new_full<RuntimeApi, Executor>(
 		})
 		.collect();
 
-	let (overseer, overseer_handler) = real_overseer(leaves, prometheus_registry.as_ref(), spawner)?;
-	let overseer_handler_clone = overseer_handler.clone();
+	let authority_discovery_service = if role.is_authority() || is_collator {
+		use sc_network::Event;
+		use futures::StreamExt;
 
-	task_manager.spawn_essential_handle().spawn_blocking("overseer", Box::pin(async move {
-		use futures::{pin_mut, select, FutureExt};
+		if authority_discovery_disabled {
+			Err("Authority discovery is mandatory for a validator.")?;
+		}
 
-		let forward = polkadot_overseer::forward_events(overseer_client, overseer_handler_clone);
+		let authority_discovery_role = if role.is_authority() {
+			authority_discovery::Role::PublishAndDiscover(
+				keystore_container.keystore(),
+			)
+		} else {
+			// don't publish our addresses when we're only a collator
+			authority_discovery::Role::Discover
+		};
+		let dht_event_stream = network.event_stream("authority-discovery")
+			.filter_map(|e| async move { match e {
+				Event::Dht(e) => Some(e),
+				_ => None,
+			}});
+		let (worker, service) = authority_discovery::new_worker_and_service(
+			client.clone(),
+			network.clone(),
+			Box::pin(dht_event_stream),
+			authority_discovery_role,
+			prometheus_registry.clone(),
+		);
 
-		let forward = forward.fuse();
-		let overseer_fut = overseer.run().fuse();
+		task_manager.spawn_handle().spawn("authority-discovery-worker", worker.run());
+		Some(service)
+	} else {
+		None
+	};
 
-		pin_mut!(overseer_fut);
-		pin_mut!(forward);
+	// we'd say let overseer_handler = authority_discovery_service.map(|authority_discovery_service|, ...),
+	// but in that case we couldn't use ? to propagate errors
+	let overseer_handler = if let Some(authority_discovery_service) = authority_discovery_service {
+		let (overseer, overseer_handler) = real_overseer(
+			leaves,
+			keystore_container.sync_keystore(),
+			overseer_client.clone(),
+			availability_config?,
+			network.clone(),
+			authority_discovery_service,
+			prometheus_registry.as_ref(),
+			spawner,
+		)?;
+		let overseer_handler_clone = overseer_handler.clone();
+
+		task_manager.spawn_essential_handle().spawn_blocking("overseer", Box::pin(async move {
+			use futures::{pin_mut, select, FutureExt};
+
+			let forward = polkadot_overseer::forward_events(overseer_client, overseer_handler_clone);
 
-		loop {
-			select! {
-				_ = forward => break,
-				_ = overseer_fut => break,
-				complete => break,
+			let forward = forward.fuse();
+			let overseer_fut = overseer.run().fuse();
+
+			pin_mut!(overseer_fut);
+			pin_mut!(forward);
+
+			loop {
```

Comment:
Why do we loop here if we end anyway the first time any future resolves? 

This `loop` can be removed

Reply:
558844d9